### PR TITLE
Core: sync committee helpers

### PR DESCRIPTION
**Pull request by terencechain**
# Description

Add sync committee helpers from `hf1` to `develop`

# Other information

These changes are already in #9123

### CODE REVIEW

**Review by rauljordan — APPROVED** [review]

**Comment by terencechain** [comment]
Can we get one more approval before merging this? @prysmaticlabs/core-team

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [reply beacon-chain/core/altair/sync_committee.go:113]: Define maxRandomByte as a constant outside of this method.

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:119]: Any risk of overflow with randomByte?

Comment (prestonvanloon) [reply beacon-chain/core/altair/sync_committee.go:119]: Move `params.BeaconConfig().MaxEffectiveBalance` outside of the for loop, define variable

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:95]: Move this after the seed call so that the method can fail faster/cheaper if helpers.Seed fails.

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:101]: params.BeaconConfig().SyncCommitteeSize is called /referenced multiple times. Define a variable, especially when it is called at the end of every for loop cycle.

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:160]:
```suggestion
	cfg := params.BeaconConfig()
	modulo := mathutil.Max(1, cfg.SyncCommitteeSize/cfg.SyncCommitteeSubnetCount/cfg.TargetAggregatorsPerSyncSubcommittee)
```

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:154]: What about empty signatures? can we check `len(sig) == expected_sig_length`?

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:143]:
```suggestion
	cfg := params.BeaconConfig()
	subCommSize := cfg.SyncCommitteeSize / cfg.SyncCommitteeSubnetCount
```

Not sure if this helps in performance or just code clarity / readability.

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:104]: Is this loop expensive? It may be worth having a context check to fail faster if the deadline has been exceeded.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:50]: Why is len set to params.BeaconConfig().SyncCommitteeSize when this depends on `indices`?

File: beacon-chain/core/altair/sync_committee.go
Comment (terencechain) [reply beacon-chain/core/altair/sync_committee.go:18]: Good point. Bad habit from old code. I'll update godoc

File: beacon-chain/core/altair/sync_committee.go
Comment (terencechain) [reply beacon-chain/core/altair/sync_committee.go:50]: `len(indices)` has to be `params.BeaconConfig().SyncCommitteeSize` but using the list length is more idiomatic. I'll change to that

File: beacon-chain/core/altair/sync_committee.go
Comment (terencechain) [reply beacon-chain/core/altair/sync_committee.go:52]: Old comment. Removed

File: beacon-chain/core/altair/sync_committee.go
Comment (terencechain) [reply beacon-chain/core/altair/sync_committee.go:119]: No risk imo. Max random byte is 8 bits. To overflow, the validator needs to have around `(1 << 55) / 1000000000 = 36M` ethers as effective balance

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:57]: With regards to my other comment, do you need to validator that len(indices) does not exceed the max committee size?

File: beacon-chain/core/altair/sync_committee.go
Comment (prestonvanloon) [inline beacon-chain/core/altair/sync_committee.go:131]: Discussed offline that randomByte is a single byte which cannot be larger than 0xFF (1<<8-1) aka maxRandomByte.

Zero overflow risk here.

**Review by prestonvanloon — APPROVED** [review]
Thanks for making those changes, looks great! Just one question, not sure if it's required or not.

File: beacon-chain/core/altair/sync_committee.go
Comment (terencechain) [reply beacon-chain/core/altair/sync_committee.go:57]: Im leaning towards no. `NextSyncCommittee` is just a converting function on top of `NextSyncCommitteeIndices`. It converts: 

`[]uint{0,1,2,3}` to `SyncCommittee{pubkeys:[]byte{a,b,c,d...`

It shouldn't have knowledge of the committee size, it just converts what `NextSyncCommitteeIndices` returns. `NextSyncCommitteeIndices` could use some other values tomorrow (`SyncCommitteeSizeHardFork100`) and `NextSyncCommittee` wouldn't care

**Pull request merged**
